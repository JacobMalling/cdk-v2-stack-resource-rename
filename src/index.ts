import * as cdk from '@aws-cdk/core';

/**
 * Interface of operation used to rename stack and its resources
 */
export interface IRenameOperation {
  /**
   * Rename method to rename stack and its resources' physical names.
   * AWS generated physical names are not changed unless StackResourceRenamer
   * is created with RenameProps{ userCustomNameOnly:false }.
   * The updated stack name or resource's name is returned.
   * @param resourceName The original resource physical name (will be empty '' for AWS generated names).
   * @param resourceType The type name of CFN resource.
   */
  rename(resourceName: string, resourceType: string): string;
}

/**
 * Properties to control rename process.
 */
export interface RenameProps {
  /**
  * Only rename user provided custom names.
  * If set to false, rename() will be invoked for all resources names with or without custom names.
  *
  * @default True
  */
  readonly userCustomNameOnly?: boolean;
  /**
  * Mapping of resourceType names to physicalName fields
  * for resources whose physicalName field donot follow
  * the regular naming conventions: `${resourceType}`+'Name'
  *
  * @default {}
  */
  readonly irregularResourceNames?: { [key: string]: string };

  /**
  * An array of Resource Types whose physical names could not be changed.
  *
  * An empty array will allow the renaming for all resources. A non-empty
  * array will apply rename operation only if the Resource type is not in
  * this array.
  * @default []
  */
  readonly excludeResourceTypes?: string[];

  /**
  * An array of Resource Types whose physical names could be updated.
  *
  * An empty array will not allow any renaming to resources. A
  * non-empty array will allow renaming only if the Resource type is in
  * this array.
  * @default []
  */
  readonly includeResourceTypes?: string[];
}

/**
 * StackResourceRenamer renames stack name and stack's subordinate resources'
 * physical names, so that a CDK stack can be used to create multiple
 * stacks in same AWS environment.
 */
export class StackResourceRenamer implements cdk.IAspect {
  /**
   * Static method to rename a stack and all its subordinate resources.
   * @param stack The stack (and all its children resources) to be renamed.
   * @param renameOper RenameOperation is used to rename
   * stack name and resources' physical names. AWS generated
   * physical names are not changed unless the "props" is set with
   * RenameProps{userCustomNameOnly:false}
   * @param props Properties are set to customize rename operations.
   */
  static rename(stack: cdk.IConstruct, renameOper: IRenameOperation, props: RenameProps = {}) {
    cdk.Aspects.of(stack).add(new StackResourceRenamer(renameOper, props));
  }

  //mapping for resources whose physical names donot follow
  //the regular naming conventions: `${resourceType}`+'Name'
  private irregularNames: { [key: string]: string } = {
    Output: 'exportName',
    ScalingPolicy: 'policyName',
    SlackChannelConfiguration: 'configurationName',
    CompositeAlarm: 'alarmName',
    SecurityGroup: 'groupName',
    DBProxy: 'dbProxyName',
  };
  private includeResTypes: string[] | undefined;
  private excludeResTypes: string[] | undefined;
  //by default, only rename user provdied custom names
  private customNameOnly = true;
  private defaultNameField = 'name';
  /**
   * Construct a new StackResourceRenamer.
   * @param renameOper RenameOperation is used to rename
   * stack name and resources' physical names. AWS generated
   * physical names are not changed unless the "props" is set with
   * RenameProps{userCustomNameOnly:false}
   * @param props Properties are set to customize rename operations.
   */
  constructor(private renameOper: IRenameOperation, props: RenameProps = {}) {
    if (props.irregularResourceNames) {
      this.irregularNames = {
        ...this.irregularNames,
        ...props.irregularResourceNames,
      };
    }
    this.excludeResTypes = props.excludeResourceTypes;
    this.includeResTypes = props.includeResourceTypes;
    if (props.userCustomNameOnly!==undefined) {
      this.customNameOnly = props.userCustomNameOnly;
    }
  }
  /**
   * Implement core.IAspect interface
   * @param node CFN resources to be renamed.
   */
  visit(node: cdk.IConstruct): void {
    if (node instanceof cdk.Stack) {
      //rename stack
      this.renameResource(node, 'Stack');
    } else {
      //rename CFN resources
      let ctorName = node.constructor.name;
      if (ctorName.startsWith('Cfn')) {
        this.renameResource(node, ctorName.substring(3));
      }
    }
  }
  /**
   * Rename a CFN resource or stack.
   * @param node CFN resource or stack.
   * @param resTypeName The type name of CFN resource.
   */
  protected renameResource(node: cdk.IConstruct, resTypeName: string) {
    //check include/exclude
    if (this.excludeResTypes && this.excludeResTypes.length > 0 &&
      this.excludeResTypes.indexOf(resTypeName) !== -1) {
      return;
    }
    if (this.includeResTypes && this.includeResTypes.length > 0 &&
      this.includeResTypes.indexOf(resTypeName) === -1) {
      return;
    }
    //find the specific "name" field for CFN resources
    let physicalName = 'name';
    if (this.irregularNames[resTypeName]) {
      physicalName = this.irregularNames[resTypeName];
    } else {
      //decapitalize regular resource names
      let [first, ...rest] = resTypeName;
      let decapName = first.toLowerCase() + rest.join('');
      physicalName = `${decapName}Name`;
    }
    if (physicalName.length === 0) {
      return;
    }
    //some names (eg. stackName, exportName) only has getter,
    //need access protected fields starting with underscore
    let underscoreName = '_' + physicalName;
    //rename
    let b = (node as any);
    if (b[physicalName] && b[physicalName].length > 0) {
      let resName = this.isTarget(b[physicalName]);
      if (isWritable(b, physicalName) && resName.ok) {
        b[physicalName] = this.renameOper.rename(resName.value, resTypeName);
      } else if (b[underscoreName] && b[underscoreName].length > 0 && isWritable(b, underscoreName)) {
        resName = this.isTarget(b[underscoreName]);
        if (resName.ok) {
          b[underscoreName] = this.renameOper.rename(resName.value, resTypeName);
        }
      }
    } else if (b[this.defaultNameField] && b[this.defaultNameField].length > 0 && isWritable(b, this.defaultNameField)) {
      let resName = this.isTarget(b[this.defaultNameField]);
      if (resName.ok) {
        b[this.defaultNameField] = this.renameOper.rename(resName.value, resTypeName);
      }
    }
  }
  /**
   * check if a resName(resource name) is a valid target for rename;
   * if valid, return correct resName and ok:true, otherwise return ok:false
   */
  isTarget(resName: any): {value: string; ok:boolean} {
    let isAWSGenerated = cdk.Token.isUnresolved(resName);
    if (this.customNameOnly && isAWSGenerated) {
      return { value: '', ok: false };
    }
    if (!this.customNameOnly && isAWSGenerated) {
      return { value: '', ok: true }; //return empty for aws generated names
    }
    return { value: resName, ok: true };
  }
}

function isWritable<T extends Object>(obj: T, key: keyof T): boolean {
  let desc: PropertyDescriptor|undefined;
  for (let o = obj; o != Object.prototype; o = Object.getPrototypeOf(o)) {
    desc = Object.getOwnPropertyDescriptor(o, key);
    if (desc !== undefined) {
      break;
    }
  }
  if (desc===undefined) {
    desc={};
  }
  return Boolean(desc.writable);
}
