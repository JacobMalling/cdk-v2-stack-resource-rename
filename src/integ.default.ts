import * as apigateway from '@aws-cdk/aws-apigateway';
import * as lambda from '@aws-cdk/aws-lambda';
import * as s3 from '@aws-cdk/aws-s3';
import * as core from '@aws-cdk/core';
import { StackResourceRenamer } from './index';

export class IntegTesting {
  readonly stack: core.Stack[]
  constructor() {
    const app = new core.App();

    const env = {
      region: process.env.CDK_DEFAULT_REGION,
      account: process.env.CDK_DEFAULT_ACCOUNT,
    };

    const stack = new core.Stack(app, 'integration-stack', { env });

    let alias = stack.node.tryGetContext('alias');
    if (alias===undefined) {
      alias='xxx';
    }
    StackResourceRenamer.rename(stack, {
      rename: (origName, _)=>{
        return origName+'-'+alias;
      },
    });

    //for integration test
    this.stack = [stack];

    //resources in stack
    const bucket = new s3.Bucket(stack, 'WidgetStore', {
      bucketName: 'widget-store-bucket',
    });

    const handler = new lambda.Function(stack, 'WidgetHandler', {
      runtime: lambda.Runtime.NODEJS_10_X,
      handler: 'index.handler',
      functionName: 'handler_func',
      code: lambda.Code.fromInline(`
                exports.handler = async (event)=> {
                    console.log('event: ',event,' env: ',process.env);
                }
            `),
      environment: {
        BUCKET: bucket.bucketName,
      },
    });

    bucket.grantReadWrite(handler);

    const api = new apigateway.RestApi(stack, 'widgets-api', {
      restApiName: 'Widget Service',
      description: 'serves widgets',
    });

    const getBind = new apigateway.LambdaIntegration(handler, {
      requestTemplates: { 'application/json': '{"statusCode": "200"}' },
    });

    api.root.addMethod('GET', getBind, {
      operationName: 'get-widget',
    });

    new core.CfnOutput(stack, 'Dns', {
      exportName: 'ApiDns',
      value: api.url,
    });
  }
}

new IntegTesting();
