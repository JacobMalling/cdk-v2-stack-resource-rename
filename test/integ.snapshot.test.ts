import { Template } from 'aws-cdk-lib/assertions';
import { IntegTesting } from './integ.default';
import { StackResourceRenamer } from '../src/index';

test('integ snapshot validation', () => {
  const integ = new IntegTesting();
  let alias = 'xxx';
  StackResourceRenamer.rename(integ.stack[0], {
    rename: (origName, _) => {
      return origName + '-' + alias;
    },
  });
  integ.stack.forEach(stack => {
    const template = Template.fromStack(stack);
    expect(template.toJSON()).toMatchSnapshot();
  });
});
