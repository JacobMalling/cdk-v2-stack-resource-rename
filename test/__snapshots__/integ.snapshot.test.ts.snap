// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integ snapshot validation 1`] = `
Object {
  "Outputs": Object {
    "Dns": Object {
      "Export": Object {
        "Name": "ApiDns-xxx",
      },
      "Value": Object {
        "Fn::Join": Array [
          "",
          Array [
            "https://",
            Object {
              "Ref": "widgetsapi0BCA0B12",
            },
            ".execute-api.",
            Object {
              "Ref": "AWS::Region",
            },
            ".",
            Object {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            Object {
              "Ref": "widgetsapiDeploymentStageprod19E45538",
            },
            "/",
          ],
        ],
      },
    },
    "widgetsapiEndpoint2B78C075": Object {
      "Value": Object {
        "Fn::Join": Array [
          "",
          Array [
            "https://",
            Object {
              "Ref": "widgetsapi0BCA0B12",
            },
            ".execute-api.",
            Object {
              "Ref": "AWS::Region",
            },
            ".",
            Object {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            Object {
              "Ref": "widgetsapiDeploymentStageprod19E45538",
            },
            "/",
          ],
        ],
      },
    },
  },
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "WidgetHandlerF8A613CA": Object {
      "DependsOn": Array [
        "WidgetHandlerServiceRoleDefaultPolicy69F8D632",
        "WidgetHandlerServiceRole455C9244",
      ],
      "Properties": Object {
        "Code": Object {
          "ZipFile": "
                exports.handler = async (event)=> {
                    console.log('event: ',event,' env: ',process.env);
                }
            ",
        },
        "Environment": Object {
          "Variables": Object {
            "BUCKET": Object {
              "Ref": "WidgetStoreDA467545",
            },
          },
        },
        "FunctionName": "handler_func-xxx",
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "WidgetHandlerServiceRole455C9244",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "WidgetHandlerServiceRole455C9244": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "WidgetHandlerServiceRoleDefaultPolicy69F8D632": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "WidgetStoreDA467545",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "WidgetStoreDA467545",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "WidgetHandlerServiceRoleDefaultPolicy69F8D632",
        "Roles": Array [
          Object {
            "Ref": "WidgetHandlerServiceRole455C9244",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "WidgetStoreDA467545": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "BucketName": "widget-store-bucket-xxx",
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "widgetsapi0BCA0B12": Object {
      "Properties": Object {
        "Description": "serves widgets",
        "Name": "Widget Service-xxx",
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "widgetsapiAccount54650C10": Object {
      "DeletionPolicy": "Retain",
      "DependsOn": Array [
        "widgetsapi0BCA0B12",
      ],
      "Properties": Object {
        "CloudWatchRoleArn": Object {
          "Fn::GetAtt": Array [
            "widgetsapiCloudWatchRole13420B98",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
      "UpdateReplacePolicy": "Retain",
    },
    "widgetsapiCloudWatchRole13420B98": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
      "UpdateReplacePolicy": "Retain",
    },
    "widgetsapiDeployment9DE3369321af3b77e33ddabbae17e4bdf0eec01a": Object {
      "DependsOn": Array [
        "widgetsapiGETFE8E3A17",
      ],
      "Properties": Object {
        "Description": "serves widgets",
        "RestApiId": Object {
          "Ref": "widgetsapi0BCA0B12",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "widgetsapiDeploymentStageprod19E45538": Object {
      "DependsOn": Array [
        "widgetsapiAccount54650C10",
      ],
      "Properties": Object {
        "DeploymentId": Object {
          "Ref": "widgetsapiDeployment9DE3369321af3b77e33ddabbae17e4bdf0eec01a",
        },
        "RestApiId": Object {
          "Ref": "widgetsapi0BCA0B12",
        },
        "StageName": "prod-xxx",
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "widgetsapiGETApiPermissionTestintegrationstackwidgetsapiA6B0DCBCGET1DB7D880": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "WidgetHandlerF8A613CA",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "widgetsapi0BCA0B12",
              },
              "/test-invoke-stage/GET/",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "widgetsapiGETApiPermissionintegrationstackwidgetsapiA6B0DCBCGET5B4A41D8": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "WidgetHandlerF8A613CA",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "widgetsapi0BCA0B12",
              },
              "/",
              Object {
                "Ref": "widgetsapiDeploymentStageprod19E45538",
              },
              "/GET/",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "widgetsapiGETFE8E3A17": Object {
      "Properties": Object {
        "AuthorizationType": "NONE",
        "HttpMethod": "GET",
        "Integration": Object {
          "IntegrationHttpMethod": "POST",
          "RequestTemplates": Object {
            "application/json": "{\\"statusCode\\": \\"200\\"}",
          },
          "Type": "AWS_PROXY",
          "Uri": Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                Object {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                Object {
                  "Fn::GetAtt": Array [
                    "WidgetHandlerF8A613CA",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "OperationName": "get-widget",
        "ResourceId": Object {
          "Fn::GetAtt": Array [
            "widgetsapi0BCA0B12",
            "RootResourceId",
          ],
        },
        "RestApiId": Object {
          "Ref": "widgetsapi0BCA0B12",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
